firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+"""SLATE is a symbolic language defining a framework for
+performing linear algebra operations on finite element
+tensors. It is similar in principle to most linear algebra
+libraries in notation.
+
+The design of SLATE was heavily influenced by UFL, and
+utilizes much of UFLs functionality for FEM-specific form
+manipulation.
+
+Unlike UFL, however, once forms are assembled into SLATE
+`Tensor` objects, one can utilize the operations defined
+in SLATE to express complicated linear algebra operations.
+(Such as the Schur-complement reduction of a block-matrix
+system)
+
+All SLATE expressions are handled by a specialized linear
+algebra compiler, the SLATE Linear Algebra Compiler (SLAC),
+which interprets SLATE expressions and produces C++ kernel
+functions to be executed within the Firedrake architecture.
+
+Written by: Thomas Gibson (t.gibson15@imperial.ac.uk)
+"""
+
+import hashlib
+import operator
+
+from slate_assertions import *
+from slate_equation import SlateEquation
+
+from ufl.algorithms.domain_analysis import canonicalize_metadata
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.algorithms.signature import compute_expression_hashdata, compute_terminal_hashdata
+from ufl.domain import join_domains, sort_domains
+from ufl.form import _sorted_integrals
+
+
+__all__ = ['CheckRestrictions', 'RemoveRestrictions',
+           'SlateIntegral', 'Tensor',
+           'Scalar', 'Vector', 'Matrix',
+           'Inverse', 'Transpose', 'UnaryOp', 'Negative', 'Positive',
+           'BinaryOp', 'TensorAdd', 'TensorSub', 'TensorMul']
+
+
+class CheckRestrictions(MultiFunction):
+    """UFL MultiFunction for enforcing cell-wise integrals to contain
+    only positive restrictions."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def negative_restrictions(self, o):
+        raise ValueError("Cell-wise integrals must contain only positive restrictions.")
+
+
+class RemoveRestrictions(MultiFunction):
+    """UFL MultiFunction for removing any restrictions on the integrals of forms."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def positive_restricted(self, o):
+        return self(o.ufl_operands[0])
+
+
+class SlateIntegral(object):
+    """An integral wrapping class that behaves like a UFL integral,
+    but doesn't have integrands in the same way. Its purpose is to
+    provide the appropriate integration domains and subdomain data."""
```

Comment:
The purpose of this class was to provide some way to access the integrals of a full SLATE expression. For example, suppose `A` `B` and `C` are `Matrix` objects that have associated `UFL` forms (assume they have appropriate dimensions). Now what are the `UFL` integrals of the following tensor: `B * A.inv * B.T + C`?

Now why do I want to access the integrals? Well take a gander at `assemble.py`. The integrals are accessed and used, exclusively I believe, for subdomain and integration domain information. So the `SlateIntegral` is a wrapper class that provides the minimal information needed for `assemble`. As the SLATE expression becomes more complex, you can imagine that the `UFL` integrals become increasingly more difficult to manage. This wrapper class could potentially simplify how integrals are used (at least at the SLATE level).

Now, note that I am currently reworking how SLATE handles integrals. You recall I mentioned my worry about how the tensor classes are performing "integral arithmetic". So after I sort out a cleaner way to handle integrals, the `SlateIntegral` may indeed become obsolete.